proto/pulumi/codegen/schema/schema.proto

```diff
@@ -0,0 +1,215 @@
+  map<string, bytes> language = 15;
+  // AN optional object to define parameterization for the package
+  optional Parameterization parameterization = 16;
+  // A list of allowed package names in addition to the name property
+  repeated string allowed_package_names = 17;
+}
+
+message Parameterization {
+  // The definition of the base provider for the package
+  optional BaseProvider base_provider = 1;
+  // The parameter for the provider
+  optional bytes parameter = 2;
+}
+
+message BaseProvider {
+  // The unqualified name of the package
+  string name = 1;
+  // The version of the package. The version must be valid semver
+  string version = 2;
+}
+
+message Meta {
+  // A regex that is used by the importer to extract a module name from the module portion of
+  // a type token. Packages that use the module format \"namespace1/namespace2/.../namespaceN\"
+  // do not need to specify a format. The regex must define one capturing group that contains
+  // the module name, which must be formatted as \"namespace1/namespace2/...namespaceN\"
+  optional string module_format = 1;
+  // Write the package to support the pack command
+  optional bool support_pack = 2;
+}
+
+// A reference to the type.
+message TypeSpec {
+  // Indicates that when used as an input, this type does not accept eventual values.
+  optional bool plain = 1;
+  oneof type {
+    // A reference to a primitive type.
+    string primitive_type = 2;
+    // A reference to an array type.
+    TypeList array_type = 3;
+    // A reference to a map type.
+    TypeMap map_type = 4;
+    // A reference to a type by its name.
+    string ref = 5;
+    // A reference to a union type.
+    UnionType union = 6;
+  }
+}
+
+message TypeList {
+  // An array of elements
+  repeated TypeSpec type = 1;
+}
+
+message TypeMap {
+  // A map of elements
+  map<string, TypeSpec> type = 1;
+}
+
+message PropertySpec {
+  // The typespec of the property
+  TypeSpec type_spec = 1;
+  // The description of the property if any. Interpreted as Markdown
+  optional string description = 2;
+  // The constant value for the property. Must be assignable to the type of the property.
+  optional Value const = 3;
+  // The default value for the property if any.  Must by assignable to the type of the property.
+  optional Value default = 4;
+  // Additional information about the property's default value if any.
+  optional DefaultInfo default_info = 5;
+  // Indicates whether a property is deprecated.
+  optional string deprecation_message = 6;
+  // Additional language-specific data about the property
+  map<string, string> language = 7;
+  // Indicates whether the property is secret.
+  optional bool secret = 8;
+  // Indicates whether the property should be replaced on changes.
+  optional bool replace_on_changes = 9;
+  // Indicates whether the provider will replace the resource when this property is changed.
+  optional bool will_replace_on_changes = 10;
+}
+
+enum Language {
+  NodeJS = 0;
```

Comment:
Oh no, do not hardcode the list of languages. Remember long term third party languages need to be able to just hook into the system.

Reply:
Should we just make this a string instead then?  I modeled this after the JSON schema, where it's an enum.

Reply:
Yes